### DISCUSSION TITLE
DataGrid AddNew

### DISCUSSION THREAD

**Discussion by CrossSlide** [MarimerLLC/csla#1717] [Questions]
Using a Windows DataGrid I need to add a new row when the user types in the empty row at the bottom of the datagrid. I get "InvalidOperationException: List item must be marked as a child object" .

I've asked this question before [here](https://github.com/MarimerLLC/cslaforum/issues/543). Rocky said I should be able to handle a datagrid event via mycollection.AddNew(). I never got back to this issue until now.

I've tried handling AddingNewItem() and InitializingNewItem() events but I keep getting the child error. I've even tried passing an child object directly but it still fails.

Does anyone have an example of how to do this?

Thanks

WPF
CSLA: v4.6

**Comment by rockfordlhotka** [comment]
It has been some time since I looked at it, but I think the [SimpleNtier](https://github.com/MarimerLLC/csla/tree/master/Samples/SimpleNTier) sample does this.

**Reply by kcabral817** [reply]
I had the same problem with Infragistics xamdatagrid.  The datagrid invokes the object class's default constructor to create a new instance for the row.  The only way to intercept the creation is to cancel the event at the datagrid and call create_child on your own.  Our you do what I did, and that's put MarkAsChild in the default constructor.  Your individual results may vary :-)

**Comment by CrossSlide** [comment]
Thanks Rocky I'll check if out...

I did find a way to do it.

I created a method in MyChild that simply creates a child object then sets e.NewItem with it in AddingNewItem().

`       private void dcGrid_AddingNewItem(object sender, AddingNewItemEventArgs e)
        {
            e.NewItem = MyChild.NewAsChild(); 
        }
`

If anyone has a better way just let me know... Thanks

**Reply by rockfordlhotka** [reply]
Are you creating the new child using the `new` keyword, or by calling `DataPortal.CreateChild`?

**Reply by CrossSlide** [reply]
`   public static MyChild NewAsChild()
        {
            return DataPortal.CreateChild<MyChild>(); 
        }`
